File: java/src/test/java/mrk/hackerrank/HackerRank02Test.java
package mrk.hackerrank;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

class HackerRank02Test {

    private static Stream<Arguments> testData() {
        return Stream.of(
                Arguments.of(1, "#"),
                Arguments.of(2, " #\n##"),
                Arguments.of(3, "  #\n ##\n###"),
                Arguments.of(4, "   #\n  ##\n ###\n####")
        );
    }

    @ParameterizedTest
    @MethodSource("testData")
    void test(int n, String expected) {
        Assertions.assertEquals(expected, new HackerRank02().staircase(n));
    }

    @ParameterizedTest
    @MethodSource("invalidInputs")
    void testInvalidInput(int n) {
        Assertions.assertThrows(IllegalArgumentException.class, () -> new HackerRank02().staircase(n));
    }

    private static Stream<Arguments> invalidInputs() {
        return Stream.of(
                Arguments.of(0),
                Arguments.of(-1),
                Arguments.of(101)
        );
    }
}

File: kotlin/src/test/kotlin/HackerRank02Test.kt
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class HackerRank02Test {
    @ParameterizedTest(name = "{0}")
    @MethodSource("hackerRank02Data")
    fun test_hackerRank02(
        name: String,
        input: Int,
        output: String,
    ) {
        assertEquals(output, hackerRank02(input))
    }

    companion object {
        @JvmStatic
        fun hackerRank02Data() =
            listOf(
                Arguments.of("n=1", 1, "#"),
                Arguments.of("n=2", 2, " #\n##"),
                Arguments.of("n=3", 3, "  #\n ##\n###"),
                Arguments.of("n=4", 4, "   #\n  ##\n ###\n####"),
            )
    }
}
